lib/filesystem.js

```diff
@@ -0,0 +1,164 @@
+// the filesystem functions exposed in this file.
+
+/**
+ * @summary Maximum I/O retries
+ * @constant
+ * @type {Number}
+ */
+const MAXIMUM_RETRIES = 10;
+
+/**
+ * @summary I/O retry base timeout, in milliseconds
+ * @constant
+ * @type {Number}
+ */
+const RETRY_BASE_TIMEOUT = 500;
+
+/**
+ * @summary Write a chunk to a file descriptor
+ * @function
+ * @public
+ *
+ * @description
+ * This function is simply a convenient wrapper around `fs.write()`.
+ *
+ * @param {Number} fd - file descriptor
+ * @param {Buffer} chunk - buffer chunk
+ * @param {Number} position - file position to write to
+ * @param {Function} callback - callback (error)
+ *
+ * @example
+ * const fd = fs.openSync('/dev/rdisk2', 'rs+');
+ *
+ * filesystem.writeChunk(fd, Buffer.allocUnsafe(65536), 0, (error) => {
+ *   if (error) {
+ *     throw error;
+ *   }
+ *
+ *   console.log('We wrote a 65536 bytes chunk at position 0');
+ * });
+ */
+exports.writeChunk = (fd, chunk, position, callback, retries = 1) => {
+  fs.write(fd, chunk, 0, chunk.length, position, (error, bytesWritten) => {
+    if (error) {
+
+      // In some faulty SDCards or SDCard readers you might randomly
+      // get EIO errors, but they usually go away after some retries.
+      if (error.code === 'EIO') {
+        if (retries > MAXIMUM_RETRIES) {
+          return callback(error);
+        }
+
+        return setTimeout(() => {
+          return exports.writeChunk(fd, chunk, position, callback, retries + 1);
+        }, RETRY_BASE_TIMEOUT * retries);
+      }
+
+      return callback(error);
+    }
+
+    if (bytesWritten !== chunk.length) {
+      return callback(new Error(`Write ${bytesWritten} bytes, ${chunk.length} expected`));
+    }
+
+    return callback();
+  });
+};
+
+/**
+ * @summary Read a chunk from a file descriptor
+ * @function
+ * @public
+ *
+ * @description
+ * This function is simply a convenient wrapper around `fs.read()`.
+ *
+ * @param {Number} fd - file descriptor
+ * @param {Number} size - chunk size to read
+ * @param {Number} position - position to read from
+ * @param {Function} callback - callback (error, buffer)
+ *
+ * @example
+ * const fd = fs.openSync('/dev/rdisk2', 'rs+');
+ *
+ * filesystem.readChunk(fd, 512, 0, (error, buffer) => {
+ *   if (error) {
+ *     throw error;
+ *   }
+ *
+ *   console.log(buffer);
+ * });
+ */
+exports.readChunk = (fd, size, position, callback, retries = 1) => {
+  fs.read(fd, Buffer.allocUnsafe(size), 0, size, position, (error, bytesRead, buffer) => {
+
+    if (error) {
+
+      // In some faulty SDCards or SDCard readers you might randomly
+      // get EIO errors, but they usually go away after some retries.
+      if (error.code === 'EIO') {
+        if (retries > MAXIMUM_RETRIES) {
+          return callback(error);
+        }
+
+        return setTimeout(() => {
+          return exports.readChunk(fd, size, position, callback, retries + 1);
+        }, RETRY_BASE_TIMEOUT * retries);
```

Comment:
Does this increasing-timeout mean that it'll actually take a totally-broken SD card 27.5 seconds before displaying an error? (seems like quite long - I wonder if either `RETRY_BASE_TIMEOUT` or `MAXIMUM_RETRIES` should be reduced?)